File: pegasus/apps/driver-app-android/app/src/main/java/com/pegasus/driver/util/BatteryGuard.kt
package com.pegasus.driver.util

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager

object BatteryGuard {
    const val WARN_THRESHOLD = 20
    const val BLOCK_THRESHOLD = 10

    enum class DepartGate { OK, WARN_LOW, BLOCK_CRITICAL }

    fun currentLevelPercent(context: Context): Int {
        val batteryStatus = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val level = batteryStatus?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
        val scale = batteryStatus?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
        if (level < 0 || scale <= 0) return 100
        return (level * 100) / scale
    }

    fun departGate(context: Context): DepartGate {
        val level = currentLevelPercent(context)
        return when {
            level < BLOCK_THRESHOLD -> DepartGate.BLOCK_CRITICAL
            level < WARN_THRESHOLD -> DepartGate.WARN_LOW
            else -> DepartGate.OK
        }
    }
}

File: pegasus/apps/driverappios/driverappios/Utilities/BatteryGuard.swift
//
//  BatteryGuard.swift
//  driverappios
//

import UIKit

enum BatteryGuard {
    static let warnThreshold = 20
    static let blockThreshold = 10

    enum DepartGate {
        case ok
        case warnLow(level: Int)
        case blockCritical(level: Int)
    }

    static var levelPercent: Int {
        let device = UIDevice.current
        device.isBatteryMonitoringEnabled = true
        let level = device.batteryLevel
        if level < 0 { return 100 }
        return Int(level * 100)
    }

    static func departGate() -> DepartGate {
        let level = levelPercent
        if level < blockThreshold { return .blockCritical(level: level) }
        if level < warnThreshold { return .warnLow(level: level) }
        return .ok
    }
}
